File: internal/commands/update_harness_test.go
package commands

// update_harness_test.go — every test redirects HOME to a t.TempDir() and
// never runs a harness-mutating command against the real user home
// directory (see docs/req/ handoff restriction for ML-6B).

import (
	"encoding/json"
	"os"
	"path/filepath"
	"strings"
	"testing"

	"github.com/kgsaran/trackfw/internal/generators"
)

func TestUpdateHarnessCmd_RunsOutsideProjectWithoutTrackfwYAML(t *testing.T) {
	home := t.TempDir()
	t.Setenv("HOME", home)
	cwd := t.TempDir()
	orig, _ := os.Getwd()
	if err := os.Chdir(cwd); err != nil {
		t.Fatal(err)
	}
	t.Cleanup(func() { _ = os.Chdir(orig) })

	cmd := newUpdateHarnessCmd()
	var out strings.Builder
	cmd.SetOut(&out)
	cmd.SetArgs([]string{})
	if err := cmd.Execute(); err != nil {
		t.Fatalf("trackfw update harness failed outside a project: %v", err)
	}
	if out.Len() == 0 {
		t.Fatal("expected a text report on stdout")
	}
}

func TestUpdateHarnessCmd_EmptyHarnessExitsZero(t *testing.T) {
	home := t.TempDir()
	t.Setenv("HOME", home)

	cmd := newUpdateHarnessCmd()
	cmd.SetOut(&strings.Builder{})
	cmd.SetArgs([]string{})
	if err := cmd.Execute(); err != nil {
		t.Fatalf("expected exit 0 (nil error) for an empty harness, got: %v", err)
	}
}

func TestUpdateHarnessCmd_UnknownTargetIsUsageError(t *testing.T) {
	home := t.TempDir()
	t.Setenv("HOME", home)

	cmd := newUpdateHarnessCmd()
	cmd.SetOut(&strings.Builder{})
	cmd.SetArgs([]string{"--targets", "not-a-real-target"})
	if err := cmd.Execute(); err == nil {
		t.Fatal("expected an error for an unknown --targets id")
	}
}

// TestUpdateHarnessCmd_JSONKeyOrderMatchesCliParityContract asserts the
// literal key order of the serialized document — not just presence of keys.
// docs/cli-parity.md pins scope, dry_run, targets, summary at the root, and
// id, state, path (message only when present, last) inside each target. This
// mirrors the barrier command's existing key-order regression coverage
// (see docs/cli-parity.md's own warning about the ML-2E gates divergence).
func TestUpdateHarnessCmd_JSONKeyOrderMatchesCliParityContract(t *testing.T) {
	home := t.TempDir()
	t.Setenv("HOME", home)

	cmd := newUpdateHarnessCmd()
	var out strings.Builder
	cmd.SetOut(&out)
	cmd.SetArgs([]string{"--json", "--targets", "claude-skill"})
	if err := cmd.Execute(); err != nil {
		t.Fatal(err)
	}

	line := strings.TrimSpace(out.String())
	wantRootOrder := []string{`"scope"`, `"dry_run"`, `"targets"`, `"summary"`}
	assertKeyOrder(t, line, wantRootOrder)

	wantTargetOrder := []string{`"id"`, `"state"`, `"path"`}
	assertKeyOrder(t, line, wantTargetOrder)

	// Decode and check shape/values too, not only key order.
	var doc struct {
		Scope   string `json:"scope"`
		DryRun  bool   `json:"dry_run"`
		Targets []struct {
			ID    string `json:"id"`
			State string `json:"state"`
			Path  string `json:"path"`
		} `json:"targets"`
		Summary struct {
			Updated int `json:"updated"`
			Skipped int `json:"skipped"`
			Missing int `json:"missing"`
			Failed  int `json:"failed"`
		} `json:"summary"`
	}
	if err := json.Unmarshal([]byte(line), &doc); err != nil {
		t.Fatalf("invalid JSON: %v\n%s", err, line)
	}
	if doc.Scope != "harness" {
		t.Fatalf("scope = %q, want harness", doc.Scope)
	}
	if len(doc.Targets) != 1 || doc.Targets[0].ID != "claude-skill" {
		t.Fatalf("unexpected targets: %+v", doc.Targets)
	}
	if doc.Targets[0].State != "missing" {
		t.Fatalf("state = %q, want missing on an empty harness", doc.Targets[0].State)
	}
	// Summary must always carry all four counters, including zeros.
	if !strings.Contains(line, `"updated":0`) || !strings.Contains(line, `"skipped":0`) ||
		!strings.Contains(line, `"missing":1`) || !strings.Contains(line, `"failed":0`) {
		t.Fatalf("summary must always emit all four counters including zeros: %s", line)
	}
	// A target with no failure must omit "message" entirely — never emit it as "".
	if strings.Contains(line, `"message"`) {
		t.Fatalf("target without a failure must omit \"message\" entirely: %s", line)
	}
}

func TestUpdateHarnessCmd_FailedTargetIncludesMessageAndExitsNonZero(t *testing.T) {
	home := t.TempDir()
	t.Setenv("HOME", home)

	// Force a failure: make the claude-skill destination a directory instead
	// of a writable file location, so os.ReadFile succeeds reading a dir
	// (fails) and any write attempt fails too.
	skillPath := generators.GlobalClaudeSkillPath(home)
	if err := os.MkdirAll(skillPath, 0755); err != nil {
		t.Fatal(err)
	}

	cmd := newUpdateHarnessCmd()
	var out strings.Builder
	cmd.SetOut(&out)
	cmd.SetArgs([]string{"--json", "--targets", "claude-skill"})
	err := cmd.Execute()
	if err == nil {
		t.Fatal("expected a non-nil error when a target fails")
	}

	line := strings.TrimSpace(out.String())
	if !strings.Contains(line, `"state":"failed"`) {
		t.Fatalf("expected state failed in JSON output: %s", line)
	}
	if !strings.Contains(line, `"message"`) {
		t.Fatalf("failed target must include a message: %s", line)
	}
}

// TestUpdateHarnessCmd_CredentialGuardClaudeInstallsViaCLI exercises the
// claude-credential-guard target through the full `trackfw update harness`
// CLI surface (not just the generators.UpdateHarness API), confirming the
// --targets/--install-missing/--json flags all thread through correctly for
// this new target.
func TestUpdateHarnessCmd_CredentialGuardClaudeInstallsViaCLI(t *testing.T) {
	home := t.TempDir()
	t.Setenv("HOME", home)

	cmd := newUpdateHarnessCmd()
	var out strings.Builder
	cmd.SetOut(&out)
	cmd.SetArgs([]string{"--json", "--targets", "claude-credential-guard", "--install-missing"})
	if err := cmd.Execute(); err != nil {
		t.Fatalf("trackfw update harness --targets claude-credential-guard --install-missing failed: %v", err)
	}

	line := strings.TrimSpace(out.String())
	var doc struct {
		Targets []struct {
			ID    string `json:"id"`
			State string `json:"state"`
			Path  string `json:"path"`
		} `json:"targets"`
	}
	if err := json.Unmarshal([]byte(line), &doc); err != nil {
		t.Fatalf("invalid JSON: %v\n%s", err, line)
	}
	if len(doc.Targets) != 1 || doc.Targets[0].ID != "claude-credential-guard" {
		t.Fatalf("unexpected targets: %+v", doc.Targets)
	}
	if doc.Targets[0].State != "updated" {
		t.Fatalf("state = %q, want updated", doc.Targets[0].State)
	}
	if doc.Targets[0].Path != "~/.claude/settings.json" {
		t.Fatalf("path = %q, want ~/.claude/settings.json", doc.Targets[0].Path)
	}

	settingsPath := filepath.Join(home, ".claude", "settings.json")
	data, err := os.ReadFile(settingsPath)
	if err != nil {
		t.Fatalf("~/.claude/settings.json was not written: %v", err)
	}
	wantScript := filepath.Join(home, ".trackfw", "scripts", "trackfw-credential-guard.sh")
	if !strings.Contains(string(data), wantScript) {
		t.Fatalf("settings.json does not reference the absolute global script path %s:\n%s", wantScript, data)
	}
}

// TestUpdateHarnessCmd_CredentialGuardCodexInstallsViaCLI mirrors
// TestUpdateHarnessCmd_CredentialGuardClaudeInstallsViaCLI for the
// codex-credential-guard target (ROADMAP-2026-08-06 Wave 2/ML-2B).
func TestUpdateHarnessCmd_CredentialGuardCodexInstallsViaCLI(t *testing.T) {
	home := t.TempDir()
	t.Setenv("HOME", home)

	cmd := newUpdateHarnessCmd()
	var out strings.Builder
	cmd.SetOut(&out)
	cmd.SetArgs([]string{"--json", "--targets", "codex-credential-guard", "--install-missing"})
	if err := cmd.Execute(); err != nil {
		t.Fatalf("trackfw update harness --targets codex-credential-guard --install-missing failed: %v", err)
	}

	line := strings.TrimSpace(out.String())
	var doc struct {
		Targets []struct {
			ID    string `json:"id"`
			State string `json:"state"`
			Path  string `json:"path"`
		} `json:"targets"`
	}
	if err := json.Unmarshal([]byte(line), &doc); err != nil {
		t.Fatalf("invalid JSON: %v\n%s", err, line)
	}
	if len(doc.Targets) != 1 || doc.Targets[0].ID != "codex-credential-guard" {
		t.Fatalf("unexpected targets: %+v", doc.Targets)
	}
	if doc.Targets[0].State != "updated" {
		t.Fatalf("state = %q, want updated", doc.Targets[0].State)
	}
	if doc.Targets[0].Path != "~/.codex/hooks.json" {
		t.Fatalf("path = %q, want ~/.codex/hooks.json", doc.Targets[0].Path)
	}

	hooksPath := filepath.Join(home, ".codex", "hooks.json")
	data, err := os.ReadFile(hooksPath)
	if err != nil {
		t.Fatalf("~/.codex/hooks.json was not written: %v", err)
	}
	wantScript := filepath.Join(home, ".trackfw", "scripts", "trackfw-credential-guard.sh")
	if !strings.Contains(string(data), wantScript) {
		t.Fatalf("hooks.json does not reference the absolute global script path %s:\n%s", wantScript, data)
	}
}

// TestUpdateHarnessCmd_CredentialGuardGeminiInstallsViaCLI mirrors
// TestUpdateHarnessCmd_CredentialGuardCodexInstallsViaCLI for the
// gemini-credential-guard target (ROADMAP-2026-08-06 Wave 2/ML-2C).
func TestUpdateHarnessCmd_CredentialGuardGeminiInstallsViaCLI(t *testing.T) {
	home := t.TempDir()
	t.Setenv("HOME", home)

	cmd := newUpdateHarnessCmd()
	var out strings.Builder
	cmd.SetOut(&out)
	cmd.SetArgs([]string{"--json", "--targets", "gemini-credential-guard", "--install-missing"})
	if err := cmd.Execute(); err != nil {
		t.Fatalf("trackfw update harness --targets gemini-credential-guard --install-missing failed: %v", err)
	}

	line := strings.TrimSpace(out.String())
	var doc struct {
		Targets []struct {
			ID    string `json:"id"`
			State string `json:"state"`
			Path  string `json:"path"`
		} `json:"targets"`
	}
	if err := json.Unmarshal([]byte(line), &doc); err != nil {
		t.Fatalf("invalid JSON: %v\n%s", err, line)
	}
	if len(doc.Targets) != 1 || doc.Targets[0].ID != "gemini-credential-guard" {
		t.Fatalf("unexpected targets: %+v", doc.Targets)
	}
	if doc.Targets[0].State != "updated" {
		t.Fatalf("state = %q, want updated", doc.Targets[0].State)
	}
	if doc.Targets[0].Path != "~/.gemini/settings.json" {
		t.Fatalf("path = %q, want ~/.gemini/settings.json", doc.Targets[0].Path)
	}

	settingsPath := filepath.Join(home, ".gemini", "settings.json")
	data, err := os.ReadFile(settingsPath)
	if err != nil {
		t.Fatalf("~/.gemini/settings.json was not written: %v", err)
	}
	wantScript := filepath.Join(home, ".trackfw", "scripts", "trackfw-credential-guard.sh")
	if !strings.Contains(string(data), wantScript) {
		t.Fatalf("settings.json does not reference the absolute global script path %s:\n%s", wantScript, data)
	}
}

// TestUpdateHarnessCmd_CredentialGuardCursorInstallsViaCLI mirrors
// TestUpdateHarnessCmd_CredentialGuardGeminiInstallsViaCLI for the
// cursor-credential-guard target (ROADMAP-2026-08-06 Wave 2/ML-2D).
func TestUpdateHarnessCmd_CredentialGuardCursorInstallsViaCLI(t *testing.T) {
	home := t.TempDir()
	t.Setenv("HOME", home)

	cmd := newUpdateHarnessCmd()
	var out strings.Builder
	cmd.SetOut(&out)
	cmd.SetArgs([]string{"--json", "--targets", "cursor-credential-guard", "--install-missing"})
	if err := cmd.Execute(); err != nil {
		t.Fatalf("trackfw update harness --targets cursor-credential-guard --install-missing failed: %v", err)
	}

	line := strings.TrimSpace(out.String())
	var doc struct {
		Targets []struct {
			ID    string `json:"id"`
			State string `json:"state"`
			Path  string `json:"path"`
		} `json:"targets"`
	}
	if err := json.Unmarshal([]byte(line), &doc); err != nil {
		t.Fatalf("invalid JSON: %v\n%s", err, line)
	}
	if len(doc.Targets) != 1 || doc.Targets[0].ID != "cursor-credential-guard" {
		t.Fatalf("unexpected targets: %+v", doc.Targets)
	}
	if doc.Targets[0].State != "updated" {
		t.Fatalf("state = %q, want updated", doc.Targets[0].State)
	}
	if doc.Targets[0].Path != "~/.cursor/hooks.json" {
		t.Fatalf("path = %q, want ~/.cursor/hooks.json", doc.Targets[0].Path)
	}

	hooksPath := filepath.Join(home, ".cursor", "hooks.json")
	data, err := os.ReadFile(hooksPath)
	if err != nil {
		t.Fatalf("~/.cursor/hooks.json was not written: %v", err)
	}
	wantScript := filepath.Join(home, ".trackfw", "scripts", "trackfw-credential-guard.sh")
	if !strings.Contains(string(data), wantScript) {
		t.Fatalf("hooks.json does not reference the absolute global script path %s:\n%s", wantScript, data)
	}
}

func assertKeyOrder(t *testing.T, doc string, keys []string) {
	t.Helper()
	var positions []int
	for _, key := range keys {
		pos := strings.Index(doc, key)
		if pos < 0 {
			t.Fatalf("expected key %s to be present in %s", key, doc)
		}
		positions = append(positions, pos)
	}
	for i := 1; i < len(positions); i++ {
		if positions[i-1] >= positions[i] {
			t.Fatalf("expected key order %v, got JSON with wrong order: %s", keys, doc)
		}
	}
}

File: npm/tests/update-harness.test.js
'use strict'

// ML-6C (ROADMAP-2026-07-29-barrier-governanca-e-autoridade-do-orquestrador)
// — see docs/cli-parity.md, "`trackfw update` vs `trackfw update harness`".
//
// EVERY test in this file redirects HOME to a scratch directory (via
// spawnSync env, never process.env.HOME mutation in-process) and NEVER
// invokes `trackfw update harness` against the real HOME.

const test = require('node:test')
const assert = require('node:assert/strict')
const fs = require('node:fs')
const os = require('node:os')
const path = require('node:path')
const { spawnSync } = require('node:child_process')

const bin = path.resolve(__dirname, '../bin/trackfw')
const { HARNESS_TARGET_IDS, claudeSkillContent } = require('../src/commands/update-harness')

function scratchHome() {
  return fs.mkdtempSync(path.join(os.tmpdir(), 'trackfw-harness-test-'))
}

function run(args, homeRoot, cwd) {
  return spawnSync(process.execPath, [bin, ...args], {
    cwd: cwd || homeRoot,
    env: { ...process.env, HOME: homeRoot },
    encoding: 'utf8',
  })
}

test('update harness never requires trackfw.yaml or a project cwd', () => {
  const homeRoot = scratchHome()
  const cwd = fs.mkdtempSync(path.join(os.tmpdir(), 'trackfw-harness-nowhere-'))
  const result = run(['update', 'harness', '--json'], homeRoot, cwd)
  assert.equal(result.status, 0, result.stderr)
  assert.doesNotThrow(() => JSON.parse(result.stdout))
})

test('update harness runs fine from inside a project directory that has its own trackfw.yaml', () => {
  const homeRoot = scratchHome()
  const projectRoot = fs.mkdtempSync(path.join(os.tmpdir(), 'trackfw-harness-inside-project-'))
  fs.writeFileSync(path.join(projectRoot, 'trackfw.yaml'), 'hooks: none\nci: none\n')
  const result = run(['update', 'harness', '--json'], homeRoot, projectRoot)
  assert.equal(result.status, 0, result.stderr)
  const doc = JSON.parse(result.stdout)
  assert.equal(doc.scope, 'harness')
})

test('an empty harness reports every declared target missing and exits 0', () => {
  const homeRoot = scratchHome()
  const result = run(['update', 'harness', '--json'], homeRoot)
  assert.equal(result.status, 0, result.stderr)
  const doc = JSON.parse(result.stdout)
  assert.equal(doc.scope, 'harness')
  assert.equal(doc.targets.length, HARNESS_TARGET_IDS.length)
  assert.deepEqual(doc.targets.map(t => t.id), HARNESS_TARGET_IDS)
  for (const t of doc.targets) assert.equal(t.state, 'missing', `${t.id} expected missing`)
  assert.equal(doc.summary.missing, HARNESS_TARGET_IDS.length)
  assert.equal(doc.summary.updated + doc.summary.skipped + doc.summary.failed, 0)
})

test('JSON document has the exact frozen key order: scope, dry_run, targets, summary', () => {
  const homeRoot = scratchHome()
  const doc = JSON.parse(run(['update', 'harness', '--json'], homeRoot).stdout)
  assert.deepEqual(Object.keys(doc), ['scope', 'dry_run', 'targets', 'summary'])
  assert.deepEqual(Object.keys(doc.summary), ['updated', 'skipped', 'missing', 'failed'])
  for (const t of doc.targets) assert.deepEqual(Object.keys(t), ['id', 'state', 'path'])
})

test('the four states appear in one document: missing, updated (via --install-missing), skipped (re-run), and failed', () => {
  const homeRoot = scratchHome()

  // updated: install-missing on a fresh target
  const installed = JSON.parse(
    run(['update', 'harness', '--json', '--install-missing', '--targets', 'claude-skill'], homeRoot).stdout
  )
  assert.equal(installed.targets[0].state, 'updated')
  assert.equal(fs.readFileSync(path.join(homeRoot, '.claude', 'skills', 'trackfw', 'SKILL.md'), 'utf8'), claudeSkillContent())

  // skipped: re-run, content already current
  const skipped = JSON.parse(
    run(['update', 'harness', '--json', '--targets', 'claude-skill'], homeRoot).stdout
  )
  assert.equal(skipped.targets[0].state, 'skipped')

  // missing: a target that was never installed
  const missing = JSON.parse(
    run(['update', 'harness', '--json', '--targets', 'codex-agents'], homeRoot).stdout
  )
  assert.equal(missing.targets[0].state, 'missing')

  // failed: block the write path (a file sits where a directory must be created)
  const homeRoot2 = scratchHome()
  fs.mkdirSync(path.join(homeRoot2, '.claude', 'skills'), { recursive: true })
  fs.writeFileSync(path.join(homeRoot2, '.claude', 'skills', 'trackfw'), 'blocking file, not a directory\n')
  const failed = run(['update', 'harness', '--json', '--install-missing', '--targets', 'claude-skill'], homeRoot2)
  assert.notEqual(failed.status, 0)
  const failedDoc = JSON.parse(failed.stdout)
  assert.equal(failedDoc.targets[0].state, 'failed')
  assert.equal(failedDoc.summary.failed, 1)
  assert.ok(failedDoc.targets[0].message, 'failed target must carry a message')
  assert.deepEqual(Object.keys(failedDoc.targets[0]), ['id', 'state', 'path', 'message'])
})

test('--dry-run never writes to HOME even with --install-missing', () => {
  const homeRoot = scratchHome()
  const before = fs.existsSync(path.join(homeRoot, '.claude'))
  const result = run(['update', 'harness', '--json', '--dry-run', '--install-missing'], homeRoot)
  assert.equal(result.status, 0, result.stderr)
  const doc = JSON.parse(result.stdout)
  assert.equal(doc.dry_run, true)
  assert.ok(doc.targets.some(t => t.state === 'updated'), 'dry-run should still predict updates')
  assert.equal(before, false)
  assert.equal(fs.existsSync(path.join(homeRoot, '.claude')), false, '--dry-run must not create ~/.claude')
})

test('--targets restricts both computation and side effects to the requested subset', () => {
  const homeRoot = scratchHome()
  const result = run(['update', 'harness', '--json', '--install-missing', '--targets', 'claude-skill'], homeRoot)
  assert.equal(result.status, 0, result.stderr)
  const doc = JSON.parse(result.stdout)
  assert.deepEqual(doc.targets.map(t => t.id), ['claude-skill'])
  // claude-agents is a distinct, always-installable target — it must not
  // have been written just because it exists in the full universe.
  assert.equal(fs.existsSync(path.join(homeRoot, '.claude', 'agents')), false)
})

test('unknown --targets id is a usage error with non-zero exit, and touches nothing', () => {
  const homeRoot = scratchHome()
  const result = run(['update', 'harness', '--targets', 'not-a-real-target'], homeRoot)
  assert.notEqual(result.status, 0)
  assert.match(result.stderr, /Unknown update target/)
  assert.equal(fs.existsSync(path.join(homeRoot, '.claude')), false)
})

test('paths are tilde-abbreviated relative to HOME, never the absolute filesystem path', () => {
  const homeRoot = scratchHome()
  const doc = JSON.parse(run(['update', 'harness', '--json'], homeRoot).stdout)
  const claudeSkill = doc.targets.find(t => t.id === 'claude-skill')
  assert.equal(claudeSkill.path, '~/.claude/skills/trackfw/SKILL.md')
  for (const t of doc.targets) assert.ok(!t.path.includes(homeRoot), `${t.id} path leaked the absolute HOME: ${t.path}`)
})

// ---------------------------------------------------------------------------
// `claude-credential-guard` — global-scope credential-guard hook wiring for
// Claude Code, ROADMAP-2026-08-06 Wave 2 ML-2A. Mirrors the Go tests in
// internal/generators/update_test.go/internal/commands/update_harness_test.go.
// ---------------------------------------------------------------------------

test('claude-credential-guard is missing without --install-missing', () => {
  const homeRoot = scratchHome()
  const doc = JSON.parse(
    run(['update', 'harness', '--json', '--targets', 'claude-credential-guard'], homeRoot).stdout
  )
  assert.equal(doc.targets[0].state, 'missing')
  assert.equal(fs.existsSync(path.join(homeRoot, '.claude', 'settings.json')), false)
})

test('claude-credential-guard installs the absolute global script path with --install-missing', () => {
  const homeRoot = scratchHome()
  const doc = JSON.parse(
    run(['update', 'harness', '--json', '--install-missing', '--targets', 'claude-credential-guard'], homeRoot).stdout
  )
  assert.equal(doc.targets[0].state, 'updated')
  assert.equal(doc.targets[0].path, '~/.claude/settings.json')

  const settingsPath = path.join(homeRoot, '.claude', 'settings.json')
  const settings = JSON.parse(fs.readFileSync(settingsPath, 'utf8'))
  const wantScript = path.join(homeRoot, '.trackfw', 'scripts', 'trackfw-credential-guard.sh')
  assert.ok(path.isAbsolute(wantScript))

  for (const event of ['PreToolUse', 'PostToolUse']) {
    const bashEntries = (settings.hooks[event] || []).filter((e) => e.matcher === 'Bash')
    assert.equal(bashEntries.length, 1)
    const commands = bashEntries[0].hooks.map((h) => h.command)
    assert.ok(commands.includes(wantScript))
  }
})

test('claude-credential-guard is idempotent', () => {
  const homeRoot = scratchHome()
  run(['update', 'harness', '--json', '--install-missing', '--targets', 'claude-credential-guard'], homeRoot)
  const settingsPath = path.join(homeRoot, '.claude', 'settings.json')
  const firstRun = fs.readFileSync(settingsPath, 'utf8')

  const doc = JSON.parse(
    run(['update', 'harness', '--json', '--install-missing', '--targets', 'claude-credential-guard'], homeRoot).stdout
  )
  assert.equal(doc.targets[0].state, 'skipped')
  const secondRun = fs.readFileSync(settingsPath, 'utf8')
  assert.equal(firstRun, secondRun)

  const settings = JSON.parse(secondRun)
  const bashEntries = settings.hooks.PreToolUse.filter((e) => e.matcher === 'Bash')
  assert.equal(bashEntries.length, 1)
})

test('claude-credential-guard --dry-run does not write', () => {
  const homeRoot = scratchHome()
  const doc = JSON.parse(
    run(
      ['update', 'harness', '--json', '--install-missing', '--dry-run', '--targets', 'claude-credential-guard'],
      homeRoot
    ).stdout
  )
  assert.equal(doc.dry_run, true)
  assert.equal(doc.targets[0].state, 'updated')
  assert.equal(fs.existsSync(path.join(homeRoot, '.claude', 'settings.json')), false)
})

test('claude-credential-guard preserves pre-existing content in ~/.claude/settings.json', () => {
  const homeRoot = scratchHome()
  const settingsPath = path.join(homeRoot, '.claude', 'settings.json')
  fs.mkdirSync(path.dirname(settingsPath), { recursive: true })
  fs.writeFileSync(
    settingsPath,
    JSON.stringify(
      {
        hooks: {
          PreToolUse: [
            {
              matcher: 'AskUserQuestion',
              hooks: [{ type: 'command', command: 'scripts/trackfw-attention-signal.sh' }],
            },
          ],
        },
        userSetting: 'keep-me',
      },
      null,
      2
    )
  )

  const doc = JSON.parse(
    run(['update', 'harness', '--json', '--install-missing', '--targets', 'claude-credential-guard'], homeRoot).stdout
  )
  assert.equal(doc.targets[0].state, 'updated')

  const settings = JSON.parse(fs.readFileSync(settingsPath, 'utf8'))
  assert.equal(settings.userSetting, 'keep-me')
  const askEntries = settings.hooks.PreToolUse.filter((e) => e.matcher === 'AskUserQuestion')
  assert.equal(askEntries.length, 1)
  assert.equal(askEntries[0].hooks[0].command, 'scripts/trackfw-attention-signal.sh')
  for (const event of ['PreToolUse', 'PostToolUse']) {
    const bashEntries = settings.hooks[event].filter((e) => e.matcher === 'Bash')
    assert.equal(bashEntries.length, 1)
  }
})

test('a project-scoped catalog install under HOME is not touched by trackfw update (project scope stays project scope)', () => {
  const homeRoot = scratchHome()
  const projectRoot = fs.mkdtempSync(path.join(os.tmpdir(), 'trackfw-harness-project-'))
  fs.writeFileSync(path.join(projectRoot, 'trackfw.yaml'), 'hooks: none\nci: none\n')
  run(['update', 'harness', '--json', '--install-missing', '--targets', 'claude-agents'], homeRoot)
  const before = fs.readFileSync(path.join(homeRoot, '.claude', 'agents', 'trackfw-architect.md'), 'utf8')
  run(['update', '--install-missing'], projectRoot, homeRoot)
  const after = fs.readFileSync(path.join(homeRoot, '.claude', 'agents', 'trackfw-architect.md'), 'utf8')
  assert.equal(before, after, 'project-scoped `update` must never touch the global harness')
})
